### PR TITLE
common: add chroot-style 'ethereal'

### PR DESCRIPTION
**Pull request by maxice8**

### CODE REVIEW

File: common/chroot-style/ethereal.sh [hunk 1]
```diff
@@ -0,0 +1,81 @@
+#!/bin/sh
```
Comment (the-maldridge) [inline common/chroot-style/ethereal.sh]: You later comment that you're using explicit bash-isms.  Make up your mind

File: common/chroot-style/ethereal.sh [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+#
+
+readonly MASTERDIR="$1"
+readonly DISTDIR="$2"
+readonly HOSTDIR="$3"
+readonly EXTRA_ARGS="$4"
+readonly CMD="$5"
+shift 5
+
+local mounts ret
```
Comment (the-maldridge) [inline common/chroot-style/ethereal.sh]: local is only valid inside a function.

**Comment by the-maldridge** [comment]
My test protocol was to use a masterdir image I prepared with docker which I have pushed for your convenience.  You should run it this way:

```
$ sudo docker run --rm -it -e XBPS_CHROOT_CMD=ethereal -v $(pwd):/hostrepo voidlinux/masterdir /bin/bash /hostrepo/xbps-src pkg go
```

It does not run as expected, my output so far is as follows:

```
=> Using `/hostrepo/hostdir/binpkgs/ethereal' as local repository.
[*] Updating `https://alpha.de.repo.voidlinux.org/current/musl/x86_64-repodata' ...
ERROR: [reposync] failed to fetch file `https://alpha.de.repo.voidlinux.org/current/musl/x86_64-repodata': Not Found
[*] Updating `https://alpha.de.repo.voidlinux.org/current/musl/nonfree/x86_64-repodata' ...
[*] Updating `https://alpha.de.repo.voidlinux.org/current/x86_64-repodata' ...
[*] Updating `https://alpha.de.repo.voidlinux.org/current/nonfree/x86_64-repodata' ...
[*] Updating `https://alpha.de.repo.voidlinux.org/current/aarch64/x86_64-repodata' ...
ERROR: [reposync] failed to fetch file `https://alpha.de.repo.voidlinux.org/current/aarch64/x86_64-repodata': Not Found
[*] Updating `https://alpha.de.repo.voidlinux.org/current/multilib/x86_64-repodata' ...
[*] Updating `https://alpha.de.repo.voidlinux.org/current/multilib/nonfree/x86_64-repodata' ...
```

**Comment by CameronNemo** [comment]
```
$ xbps-install -S podman fuse-overlayfs slirp4netns
$ podman pull docker://cameronnemo/voidlinux-9203
$ podman run --rm -it -e XBPS_CHROOT_CMD=ethereal -v $(pwd):/hostrepo docker://cameronnemo/voidlinux-9203 --security-opt apparmor=unconfined /bin/bash /hostrepo/xbps-src pkg go
```

**Comment by Gottox** [comment]
What doesn't work with the current chroot_styles and can we change them to make them work in a container?

I'm interested in *less special cases* instead of more for xbps-src.

**Comment by the-maldridge** [comment]
Fundamentally they cannot be adopted and the need for a new style extends from this.  In all of our existing chroot styles we need either setuid binaries, or the permissions to manipulate namespaces.  These are unfortunately not things we can do in commercially provided docker based CI.

The idea is that the ethereal build style supports running inside a docker container which can run in an unprivileged mode, something that none of our build styles are capable of supporting without a configuration step on the host.  Given our earlier run in with support departments, I don't think its likely that we'll be given the ability to make arbitrary configuration changes to support our unnecessarily privileged build process.

**Comment by maxice8** [comment]
@CameronNemo 

```
$ podman pull docker://cameronnemo/voidlinux-9203
ERRO[0000] No subuid ranges found for user "terran" in /etc/subuid
```

**Comment by maxice8** [comment]
```
podman run --rm -it -e XBPS_CHROOT_CMD=ethereal -v (pwd):/hostrepo docker://cameronnemo/voidlinux-9203 --security-opt apparmor=unconfined /bin/bash /hostrepo/xbps-src pkg base-files
ERRO[0000] error unmounting /home/terran/var/share/containers/storage/overlay/787fa72aba384bfe0efee0ac5f8439b80def21d4c02fff2b798298e61ab7e915/merged: invalid argument
ERRO[0000] unable to remove container 548a0aab067a5a78014c21a0ea202de88ca502914b42e9ba744c9be253e93302 after failing to start and attach to it
Error: error mounting storage for container 548a0aab067a5a78014c21a0ea202de88ca502914b42e9ba744c9be253e93302: error creating overlay mount to /home/terran/var/share/containers/storage/overlay/787fa72aba384bfe0efee0ac5f8439b80def21d4c02fff2b798298e61ab7e915/merged: exit status 1
```

@CameronNemo

**Comment by CameronNemo** [comment]
This works for me:

```
$ podman run --rm --security-opt apparmor=unconfined -it -e XBPS_CHROOT_CMD=ethereal -v (pwd):/void-packages docker://cameronnemo/voidlinux-9203 /bin/bash /void-packages/xbps-src pkg PKGNAME
```

Well at least for building base-files and podman. Tried to build XBPS and got:

```
=> xbps-0.53_8: running do_configure ...
Configuring xbps for ... linux
Using compiler cc
Building with debugging symbols.
Checking if cc supports -Wl,--no-as-needed ... no.
Checking if cc supports -Wall ... no.
Checking if cc supports -Wextra ... no.
Checking if cc supports -Werror ... no.
Checking if cc supports -Wshadow ... no.
Checking if cc supports -Wformat=2 ... no.
Checking if cc supports -Wmissing-prototypes ... no.
Checking if cc supports -Wmissing-declarations ... no.
Checking if cc supports -Wnested-externs ... no.
Checking if cc supports -Wvla ... no.
Checking if cc supports -Wno-overlength-strings ... no.
Checking if cc supports -Wunsafe-loop-optimizations ... no.
Checking if cc supports -Wundef ... no.
Checking if cc supports -Wsign-compare ... no.
Checking if cc supports -Wmissing-include-dirs ... no.
Checking if cc supports -Wold-style-definition ... no.
Checking if cc supports -Winit-self ... no.
Checking if cc supports -Wredundant-decls ... no.
Checking if cc supports -Wfloat-equal ... no.
Checking if cc supports -Wmissing-noreturn ... no.
Checking if cc supports -Wcast-align ... no.
Checking if cc supports -Wcast-qual ... no.
Checking if cc supports -Wpointer-arith ... no.
Checking if cc supports -Wcomment ... no.
Checking if cc supports -Wdeclaration-after-statement ... no.
Checking if cc supports -Wwrite-strings ... no.
Checking if cc supports -Wstack-protector ... no.
Checking if cc supports -fPIC ... no.
Checking if cc supports -finline-functions ... no.
Checking if cc supports -fstack-protector-strong ... no.
Checking if cc supports -fstack-protector ... no.
Checking if cc supports -fvisibility=default ... no.
Checking if cc supports -Wl,--export-dynamic ... no.
Checking if cc supports -rdynamic ... no.
Checking if cc supports -Wl,-z,relro,-z,now ... no.
Checking if cc supports -fPIE ... no.
Checking if cc supports -std=c99 ... no.
ERRROR: A compatible C99 compiler is required, exiting...
```

**Comment by maxice8** [comment]
@CameronNemo Sorry i only tested on docker.

Creating the container from @the-maldridge on the first comment and doing:

```
sudo docker run --rm -it -e XBPS_CHROOT_CMD=ethereal -v $(pwd):/hostrepo voidlinux/masterdir /bin/bash -c 'hostrepo/xbps-src hoststrap ; hostrepo/xbps-src pkg xbps'
```

**Comment by the-maldridge** [comment]
LGTM to me.  I've just confirmed a local build for the x86_64 target, I did not test cross targets due to now having the cross masterdirs available at the moment.  @Vaelatern for a second lookover and then I can look into pushing masterdirs to our docker hub.

**Comment by Vaelatern** [comment]
Should hoststrap also -Su the "masterdir"?

**Comment by maxice8** [comment]
On Mon, Mar 04, 2019 at 08:02:27PM -0800, Toyam Cox wrote:
> Should hoststrap also -Su the "masterdir"?
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/void-linux/void-packages/pull/9203#issuecomment-469528492

It can be done.

**Comment by bobertlo** [comment]
I found that `common/chroot-style/ethereal.sh` needs `#!/bin/sh` on musl docker hosts.

I made a musl masterdir docker image with:
```
sudo XBPS_ARCH=x86_64-musl xbps-install -y -r $(pwd)/masterdir-tmp/ -R http://mirrors.servercentral.com/voidlinux/current/musl -S base-chroot git
sudo tar pvcf masterdir-musl.tar -C masterdir-tmp/ -c .
```
I copied it to a glibc box to import, because I've never had any luck with docker on musl.
I also uploaded to docker hub under `bobertlo/void-masterdir-musl` if anyone is interested.

I built a lot of packages natively on a glibc host, and a few on musl. I also tested a few cross compiled packages on both architectures, and a couple cross glibc->musl. Mostly random packages, including gcc a couple different ways.

**Comment by maxice8** [comment]
On Mon, Mar 04, 2019 at 11:04:46PM -0800, Robert Lowry wrote:
> I found that `common/chroot-style/ethereal.sh` needs `#!/bin/sh` on musl docker hosts.
> 
> I made a musl masterdir docker image with:
> ```
> sudo XBPS_ARCH=x86_64-musl xbps-install -y -r $(pwd)/masterdir-tmp/ -R http://mirrors.servercentral.com/voidlinux/current/musl -S base-chroot git
> sudo tar pvcf masterdir-musl.tar -C masterdir-tmp/ -c .
> ```
> I copied it to a glibc box to import, because I've never had any luck with docker on musl.
> I also uploaded to docker hub under `bobertlo/void-masterdir-musl` if anyone is interested.
> 
> I built a lot of packages natively on a glibc host, and a few on musl. I also tested a few cross compiled packages on both architectures, and a couple cross glibc->musl. Mostly random packages, including gcc a couple different ways.
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/void-linux/void-packages/pull/9203#issuecomment-469564694

Added /bin/bash since we use stuff like OLDPWD

**Comment by maxice8** [comment]
scratch the above i made the script POSIX sh so /bin/sh can be used safely

File: common/xbps-src/shutils/common.sh [hunk 3]
```diff
@@ -386,7 +386,7 @@ setup_pkg() {
     esac
 
     # Check if base-chroot is already installed.
-    if [ -z "$bootstrap" -a "z$show_problems" != "zignore-problems" ]; then
+    if [ -z "$bootstrap" -a "z$show_problems" != "zignore-problems" -a "$XBPS_CHROOT_CMD" != "ethereal" ]; then
```
Comment (maxice8) [inline common/xbps-src/shutils/common.sh]: @the-maldridge I have removed this check, the side-effect is that we now require all images to be docker containers to have base-chroot installed.

File: xbps-src [hunk 4]
```diff
@@ -565,7 +570,7 @@ if [ -n "${NO_COLOR+x}" ]; then
 fi
 
 chroot_check() {
-    if [ -f $XBPS_MASTERDIR/.xbps_chroot_init ]; then
+    if [ -f $XBPS_MASTERDIR/.xbps_chroot_init  ]; then
```
Comment (Vaelatern) [inline xbps-src]: Stray patch line?

Comment (maxice8) [reply xbps-src]: yes of previous attempt

**Comment by maxice8** [comment]
Ok, if we can get people to test i think this is good to go.

**Comment by maxice8** [comment]
```
sudo docker run --rm -it \
       -e XBPS_CHROOT_CMD=ethereal \
       -e XBPS_ALLOW_CHROOT_BREAKOUT=yes \
      -v (pwd):/hostrepo voidlinux/masterdir \
      /bin/bash -c 'rm -f hostrepo/masterdir ; ln -sf / hostrepo/masterdir && hostrepo/xbps-src pkg xbps'
```

**Comment by Vaelatern** [comment]
:shipit: We'll fix any problems later.

**Comment by maxice8** [comment]
On Tue, Mar 05, 2019 at 12:31:38PM -0800, Toyam Cox wrote:
> :shipit: We'll fix any problems later.
> 
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/void-linux/void-packages/pull/9203#issuecomment-469845844

This ship won't sail until i finish beating Travis-ci up

File: common/chroot-style/ethereal.sh [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+#
+# This chroot script uses symlinks to emulate being in a chroot,
+# it also symlinks the root directory to the masterdir and assumes
+# base-chroot is installed in the host system.
+#
+# It will damage your host system, only use it in disposable
+# containers.
+#
+# Note that this assumes the following.
+# The xbps-src git checkout not is mounted at /void-packages
```
Comment (the-maldridge) [inline common/chroot-style/ethereal.sh]: is not, not not is.

File: common/chroot-style/ethereal.sh [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+readonly MASTERDIR="$1"
+readonly DISTDIR="$2"
+readonly HOSTDIR="$3"
+readonly EXTRA_ARGS="$4"
+readonly CMD="$5"
+shift 5
+
+if [ -z "$MASTERDIR" -o -z "$DISTDIR" ]; then
+	echo "$0 MASTERDIR/DISTDIR not set"
```
Comment (the-maldridge) [inline common/chroot-style/ethereal.sh]: Perhaps a slightly more helpful message here?  Something about linking the root.

File: common/chroot-style/ethereal.sh [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+	fake_mount "$from" "$to"
+
+	mounts="${mounts} $to"
+done
+
+# Store current directory for returning later
+OLDPWD="$(pwd)"
+
+# To give the illusion we entered the chroot, cd to /
+cd /
```
Comment (the-maldridge) [inline common/chroot-style/ethereal.sh]: This should probably have an || error case to handle the unlikely event that the cd fails.

File: common/chroot-style/ethereal.sh [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+# Start with `env` so our environment var's stay the same
+env IN_CHROOT=1 $CMD $@
+
+# Store return of the command we care about
+ret="$?"
+
+# Return to OLDPWD
+cd "${OLDPWD}"
+
+if [ -n "$HOSTDIR" ]; then
```
Comment (the-maldridge) [inline common/chroot-style/ethereal.sh]: Please comment why this is done.

File: common/chroot-style/ethereal.sh [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+	rm -f "$MASTERDIR"/host
+	mkdir -p "$MASTERDIR"/host
+fi
+
+for m in $mounts; do
```
Comment (the-maldridge) [inline common/chroot-style/ethereal.sh]: Same here, please comment.

File: common/chroot-style/ethereal.sh (hunk 6 above)
Comment (maxice8) [reply common/chroot-style/ethereal.sh]: That is an error for all chroot-styles, it means xbps-src somehow passed something wrong.

File: common/chroot-style/ethereal.sh [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+#
+# It will damage your host system, only use it in disposable
+# containers.
+#
+# The supported way to make use of thie chroot-style is to create
+# a root filesystem that has base-chroot and git installed and 
+# have it inside a container engine like Docker.
+#
+# Docker example:
+# $ tmp/image
```
Comment (Chocimier) [inline common/chroot-style/ethereal.sh]: is it `cd tmp/image` ?

**Comment by the-eater** [comment]
Oh, would've been fun if someone pinged me about this :eyes: at least nice to see that @maxice8 did have the energy to go through the endless bike shedding

**Pull request merged**